File: packages/scheduling-engine/src/analysis/cap-analysis.ts
// ═══════════════════════════════════════════════════════════
//  INCOMPOL PLAN — Capacity Analysis
//  Per-machine per-day load computation
//  Extracted from NikufraEngine.tsx capAnalysis()
// ═══════════════════════════════════════════════════════════

import type { Block, DayLoad } from '../types/blocks.js';
import type { EMachine } from '../types/engine.js';

/**
 * Compute per-machine per-day load from scheduled blocks.
 *
 * Returns: Record<machineId, DayLoad[]> where each DayLoad contains
 * production minutes, setup minutes, ops count, pieces, and blocked count.
 */
export function capAnalysis(
  blocks: Block[],
  machines: EMachine[],
  nDays?: number,
): Record<string, DayLoad[]> {
  const cNDays = nDays ?? (blocks.length > 0 ? Math.max(...blocks.map((b) => b.dayIdx)) + 1 : 0);
  const cap: Record<string, DayLoad[]> = {};

  machines.forEach((m) => {
    cap[m.id] = Array.from({ length: cNDays }, () => ({
      prod: 0,
      setup: 0,
      ops: 0,
      pcs: 0,
      blk: 0,
    }));
  });

  blocks.forEach((b) => {
    if (!cap[b.machineId]) return;
    const dc = cap[b.machineId][b.dayIdx];
    if (b.type === 'blocked') {
      dc.blk++;
      return;
    }
    dc.prod += b.endMin - b.startMin;
    if (b.setupS != null && b.setupE != null) dc.setup += b.setupE - b.setupS;
    dc.ops++;
    // Twin co-production: count total physical pieces from all outputs
    dc.pcs += b.isTwinProduction && b.outputs ? b.outputs.reduce((s, o) => s + o.qty, 0) : b.qty;
  });

  return cap;
}

File: apps/backend/src/domain/scheduling/analysis/cap_analysis.py
"""Capacity analysis — per-machine per-day load computation."""

from __future__ import annotations

from ..types import Block, EMachine


def cap_analysis(
    blocks: list[Block],
    machines: list[EMachine],
) -> dict[str, list[dict[str, int]]]:
    """Compute per-machine per-day load from blocks.

    Returns {machine_id: [{prod, setup}, ...]} indexed by day.
    """
    n_days = 0
    for b in blocks:
        if b.day_idx + 1 > n_days:
            n_days = b.day_idx + 1

    result: dict[str, list[dict[str, int]]] = {}
    for m in machines:
        days = [{"prod": 0, "setup": 0} for _ in range(n_days)]
        for b in blocks:
            if b.machine_id != m.id or b.day_idx < 0 or b.day_idx >= n_days:
                continue
            days[b.day_idx]["prod"] += b.prod_min
            days[b.day_idx]["setup"] += b.setup_min
        result[m.id] = days
    return result
